internal/config/tokens.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	task.FromContext(monitorCtx).Run(func(taskCtx context.Context) {
+		taskCtx, cancelTask := context.WithCancel(taskCtx)
+
+		var m sync.Mutex
+		var wg sync.WaitGroup
+
+		wg.Add(2)
+
+		if t.FromConfigFile != "" {
+			log.Debugf("monitoring tokens at %s", t.FromConfigFile)
+		} else {
+			log.Debug("monitoring tokens in memory")
+		}
+
+		go monitorConfigTokenChanges(taskCtx, &m, t, wg.Done)
+		go keepConfigTokensFresh(taskCtx, &m, t, uucb, wg.Done)
+
+		// shut down when the task manager is shutting down or when the
+		// ctx passed into MonitorTokens is cancelled.
+		select {
+		case <-taskCtx.Done():
+		case <-monitorCtx.Done():
+		}
+
+		log.Debug("done monitoring tokens")
+		cancelTask()
+		wg.Wait()
+	})
+}
+
+// monitorConfigTokenChanges watches for token changes in the config file. This can
+// happen if a foreground process updates the config file while the agent is
+// running.
+func monitorConfigTokenChanges(ctx context.Context, m *sync.Mutex, t *tokens.Tokens, done func()) error {
+	defer done()
+
+	fromConfigFile := tokensConfigFile(t, m)
+	if fromConfigFile == "" {
+		return nil
+	}
+
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			currentStr, err := ReadAccessToken(t.FromConfigFile)
+			if err != nil {
+				return err
+			}
+
+			current := tokens.Parse(currentStr)
+			current.FromConfigFile = fromConfigFile
+
+			m.Lock()
+			if t.All() != current.All() {
+
+				t.Replace(current)
+			}
+			m.Unlock()
+		}
+	}
+}
+
+// keepConfigTokensFresh periodically updates our tokens and syncs those to the config
+// file.
+func keepConfigTokensFresh(ctx context.Context, m *sync.Mutex, t *tokens.Tokens, uucb UserURLCallback, done func()) error {
+	defer done()
+
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+
+	logger := logger.FromContext(ctx)
+	fromConfigFile := tokensConfigFile(t, m)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			beforeUpdate := t.All()
+			localCopy := tokens.Parse(beforeUpdate)
+
+			updated1, err := fetchOrgTokens(ctx, localCopy)
+			if err != nil {
+				logger.Debugf("failed to fetch missing tokens for user: %s", err)
+				// don't continue. might have been partial success
+			}
+
+			updated2, err := refreshDischargeTokens(ctx, localCopy, uucb)
+			if err != nil {
+				logger.Debugf("failed to update discharge tokens: %s", err)
+				// don't continue. might have been partial success
+			}
+
+			if !updated2 && !updated1 {
+				continue
+			}
+
+			m.Lock()
+			// don't clobber config file if it changed out from under us. the
+			// consequences of a race here (agent and foreground command both
+			// fetching updates simultaneously) are low, so don't bother with an
+			// extra lock file.
+			if t.All() == beforeUpdate {
+				t.Replace(localCopy)
+
+				if fromConfigFile != "" {
+					if err := SetAccessToken(fromConfigFile, t.All()); err != nil {
```

Comment:
is `fromConfigFile` here still valid after `t.Replace` above?

Reply:
It'll be the same value.